### DISCUSSION TITLE
How to make autocomplete manually triggered?

### DISCUSSION THREAD

**Discussion by brokenthorn** [NvChad/NvChad#2608] [Q&A] [answered]
Whenever I am writing code but most often happens with comments, there are situations where the autocomplete popup shows up with some match, and while writing the comment or code in the speed of it, NvChad totally messes up what I am trying to write, as it is either inserting  a huge snippet or overwriting the word I am trying to write with something else, and all I'm doing is just simply writing.

I am not looking to disable the autocomplete popup, as I find it useful to see suggestions, but I want to have to press TAB first to actually select into it an item that I want to autocomplete, and then press Enter like usual to activate the completion, but I do not want the completion to ever activate by itself.

I have been looking into this but I am quite new to using nvim, just a few months. I believe it's the cmp plugin that I need to configure, and I need to disable the preselection to be None, as referenced here: https://github.com/hrsh7th/nvim-cmp/discussions/1411 but I am not sure how to configure cmp correctly with NvChad.

Can anyone help and is there any chance that this might become the default configuration? It's practically the default with most other editors.

Thanks!

#2193 - completely disables autocomplete popup and you have to press Ctrl-Space to trigger it, which I do not want. Two less keys to press this way.
#1667 - is also not what I want and the solutions proposed there are not very idiomatic.

**Comment by brokenthorn** [comment]
I seem to have found the solution, which is what is proposed in #2193, except it does not disable the popup like it says in the ticket, but only disables cmp from automatically performing completion with the first item (which it used to automatically select) when pressing space, which is exactly what I wanted!

```lua
local plugins = {
  -- other plugins...
  {
    "hrsh7th/nvim-cmp",
    opts = {
      completion = {
        autocomplete = false,
      },
    },
  },
  -- other plugins...
}
```

**Comment by brokenthorn** [comment]
Correction, that did not fix it. Here is a demo:
![SV8I1KAXF5](https://github.com/NvChad/NvChad/assets/2119004/3531c729-a4c2-4392-bae6-a51572750afe)

I think there might be two fixes: either change the autocomplete key, or disable preselecting of the first item.

I would prefer the second, as I like to be able to trigger with Enter.

**Reply by siduck** [reply]
@brokenthorn there's an option for the 2nd one, google up & read cmp docs, and seek help in our discord server as its more active than here

**Comment by brokenthorn** [comment, marked as answer]
Finally did it!

```lua
  {
    "hrsh7th/nvim-cmp",
    opts = {
      mapping = {
        ["<CR>"] = cmp.mapping.confirm {
          behavior = cmp.ConfirmBehavior.Insert,
          -- when true, auto-selects the first item if nothing was selected,
          -- making noselect below not take effect.
          select = false,
        },
      },
      -- adding noselect compared to default, to prevent autocomplete when typing,
      -- and this is actually nvim-cmp defaults, but NvChad overrides this.
      completion = {
        completeopt = "menu,menuone,noselect",
      },
      -- for LSPs that (re)enable this:
      preselect = cmp.PreselectMode.None,
    },
  },
```

**Reply by siduck** [reply]
@brokenthorn  your method removes all of lazyloading of cmp by calling its module directly into opts table, instead of that we make opts a function

```lua

  {
    "hrsh7th/nvim-cmp",
    opts = function()
      local config = require "plugins.configs.cmp"
      local cmp = require "cmp"

      config.mapping["<CR>"] = cmp.mapping.confirm {
        behavior = cmp.ConfirmBehavior.Insert,
        select = false,
      }

      config.completion = {
        completeopt = "menu,menuone,noselect",
      }

      config.preselect = cmp.PreselectMode.None

      return config
    end,
  }
  ```

**Reply by brokenthorn** [reply]
Hmm, I thought opts was merged with the default config and passed to setup().

**Reply by brokenthorn** [reply]
Thanks so much!

**Reply by siduck** [reply]
@brokenthorn opts does merge, but the thing is that you ran cmp module directly, which will break lazyloading of cmp, so we better wrap it in a function!

**Reply by brokenthorn** [reply]
I'm not sure where I was running it directly or what that means. Please can you explain?

**Reply by brokenthorn** [reply]
> @brokenthorn  your method removes all of lazyloading of cmp by calling its module directly into opts table, instead of that we make opts a function
> 
> ```lua
> 
>   {
>     "hrsh7th/nvim-cmp",
>     opts = function()
>       local config = require "plugins.configs.cmp"
>       local cmp = require "cmp"
> 
>       config.mapping["<CR>"] = cmp.mapping.confirm {
>         behavior = cmp.ConfirmBehavior.Insert,
>         select = false,
>       }
> 
>       config.completion = {
>         completeopt = "menu,menuone,noselect",
>       }
> 
>       config.preselect = cmp.PreselectMode.None
> 
>       return config
>     end,
>   }
>   ```

Worked perfectly. Thank you.

**Reply by siduck** [reply]
yes you ran it directly, if a module is exposed outside a function then nvim automatically executes it, but if you wrap it in a function then nvim doesnt execute it by default, it executes it when the function call happens

**Reply by clockworkpc** [reply]
Works for me too!

**Comment by chuck-r** [comment]
I just recently started using Nvim/NvChad. The instructions here didn't seem to work for me. However, I did find that this works:

~/.config/lua/nvchad/cmp.lua:
```lua
local cmp = require "cmp"

local options = {
  mapping = {
    ["<CR>"] = cmp.mapping.confirm {
      behavior = cmp.ConfirmBehavior.Insert,
      select = false,
    },
  },

  completion = {
    completeopt = "menu,menuone,noselect",
  },

  preselect = cmp.PreselectMode.None
}

return options
```

The file ~/.local/share/nvim/lazy/NvChad/lua/nvchad/configs/cmp.lua ends with a call to `return vim.tbl_deep_extend("force", options, require "nvchad.cmp")`. So, by creating the "nvchad.cmp" file, the defaults can be overwritten.

Edit: It appears this change was made [2 weeks ago](https://github.com/NvChad/NvChad/commit/7f055761ba0d94a562ead5ff85352c4c67b51f67#diff-8c63048840480d19491c64d72fb0e8a7f1be269c9a57e7143a44ff9461cf3096R57).

**Reply by siduck** [reply]
@chuck-r  that is very wrong, because `nvchad/cmp` module is literally in our ui repo. 

You are advised to use `configs/cmp.lua` or some other name. And dont forget to wrap it in a function , in your plugins file or else its gonna break lazyloading...


```lua
{ 
         "blabla cmp",
          opts =   function()
                  return require "configs.cmp"
           end
}

**Reply by chuck-r** [reply]
I see what you mean about `nvchad.cmp` being in the repo 😅. I'm also surprised that it "just worked". I wonder how the module search rules work in a case where a file is found in more than one place? I must say, as a new user of Nvim, it's hard to follow how the module search paths work, exactly. Somehow, `require "nvchad.cmp"` locates a file in `~/.local/share/nvim/lazy/ui/lua/nvchad/cmp`? Very strange.

After much trial and error, here is my final configuration. I'm including this here since, as a newbie to Nvim/NvChad, I found it a tad hard to follow @brokenthorn's example since a file name isn't given. Of course, it could all be done within `plugins/init.lua`, but storing the actual configuration in `configs/cmp.lua` nicely modularizes the configuration like other default plugins.

~/.config/nvim/lua/plugins/init.lua
```lua
[...]
{
  "hrsh7th/nvim-cmp",
  opts = function()
    return require "configs.cmp"
  end
},
[...]
```

~/.config/nvim/lua/configs/cmp.lua
```lua
local config = require "nvchad.configs.cmp"
local cmp = require "cmp"

config.mapping["<CR>"] = cmp.mapping.confirm {
  behavior = cmp.ConfirmBehavior.Insert,
  select = false,
}

config.completion = {
  completeopt = "menu,menuone,noselect"
}

config.preselect = cmp.PreselectMode.None

return config
```

**Reply by cyklee** [reply]
Changing `plugin.configs.cmp` [shown here](https://github.com/NvChad/NvChad/discussions/2608#discussioncomment-8161724) to `nvchad.config.cmp` used here worked for me.
